### ISSUE TITLE
TypeHandler are ignored by spring-data-mybatis

### ISSUE THREAD

**Issue by kopax** [easybest/spring-data-mybatis#33]
`MANAGER.ROLES` columns looks like `1,2,3`

Each ID represent a Role.

Role is an ENUM class.

I have a typeHandler to retrieve and set my roles but it seems ignored by spring-data-mybatis 

```java
package com.typeHandler;

import com.domain.Role;
import com.common.store.UniqueStore;
import org.apache.commons.lang3.StringUtils;
import org.apache.ibatis.type.BaseTypeHandler;
import org.apache.ibatis.type.JdbcType;

import java.sql.CallableStatement;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class RoleListTypeHandler<E extends Enum<Role>> extends BaseTypeHandler<List<Role>> {

	@Override
	public void setNonNullParameter(PreparedStatement ps, int i, List<Role> roleList, JdbcType jdbcType) throws SQLException {
		ArrayList<String> roles = new ArrayList<>();
		for (Role role: roleList){
			String roleId = String.valueOf(role.getValue());
			// add role only if not already present
			if( !roles.contains(roleId) )
			{
				roles.add(roleId);
			}
		}
		ps.setString(i, StringUtils.join(roles, UniqueStore.COMMA_DELIMITER));
	}

	@Override
	public List<Role> getNullableResult(ResultSet rs, String columnName) throws SQLException {
		String roles = rs.getString(columnName);
		return getRoleListFromResultString(roles);
	}

	@Override
	public List<Role> getNullableResult(ResultSet rs, int columnIndex) throws SQLException {
		return getRoleListFromResultString(rs.getString(columnIndex));
	}

	@Override
	public List<Role> getNullableResult(CallableStatement cs, int columnIndex) throws SQLException {
		return getRoleListFromResultString(cs.getString(columnIndex));
	}

	private List<Role> getRoleListFromResultString(String result)
	{
		ArrayList<Role> roleList = new ArrayList();
		List<String> roles = Arrays.asList(result.split("\\s*,\\s*"));

		for (String role : roles)
		{
			roleList.add( Role.getRoleFromValue(Integer.parseInt(role)) );
		}
		return roleList;
	}
}
```

And this is my class Role.java : 

```java
public enum Role {

	ROLE_USER(1),
	ROLE_MANAGER(2),
	ROLE_ADMIN(3),
	ROLE_CLIENT(4),
	ROLE_TRUSTED_CLIENT(5);

	private int value;

	Role(final int value) {
		this.value = value;
	}

	public int getValue() {
		return value;
	}

	public void setValue(int value) {
		this.value = value;
	}

	/**
	 * From int to Role
	 *
	 * @param value
	 * @return
	 */
	public static Role getRoleFromValue(int value) {
		Role result = null;
		for (Role s : Role.values()) {
			if (s.getValue() == value) {
				result = s;
			}
		}
		if (result == null) {
			throw new IllegalArgumentException("No exist the Role with that value");
		}
		return result;
	}
}
```

My Mybatis mapper has the following resultMap : 

```xml
<!DOCTYPE mapper PUBLIC "-//mybatis.org//DTD Mapper 3.0//EN" "http://mybatis.org/dtd/mybatis-3-mapper.dtd">
<mapper namespace="com.repository.ManagerRepository">
    <!--<cache />-->

    <resultMap id="managerResultMap" type="com.domain.Manager">
        <id property="id" column="id"/>
        <result property="login" column="login"/>
        <result property="roleList" column="roleList" typeHandler="com.typeHandler.RoleListTypeHandler" />
    </resultMap>


    <sql id="getManagerSelect">
        SELECT
            manager.`ID` as 'id',
            manager.`LOGIN` as 'login'
            manager.`ROLES` as 'roleList'
    </sql>

    <sql id="getManagerFrom">
        FROM
          `MANAGER` manager
    </sql>

    <select id="count" resultType="java.lang.Long" parameterType="Integer">
        SELECT COUNT(ID)
        <include refid="getManagerFrom"/>
    </select>

    <select id="exists" resultType="java.lang.Boolean" parameterType="Integer">
        SELECT COUNT(ID)
        <include refid="getManagerFrom"/>
        WHERE
        ID = #{id};
    </select>


    <select id="find" resultMap="managerResultMap" parameterType="Integer">
        <include refid="getManagerSelect"/>
        <include refid="getManagerFrom"/>
        <if test="pk">
            WHERE manager.`ID` = #{pk.value}
        </if>
    </select>
</mapper>
```

I have added this to my configuration

	@Bean
	public SqlSessionFactoryBean sqlSessionFactory() throws Exception {
		final SqlSessionFactoryBean sessionFactory = new SqlSessionFactoryBean();
		sessionFactory.setDataSource(dataSource());
		sessionFactory.setTypeHandlers(new TypeHandler[] {
			new DateTimeTypeHandler(),
			new RoleListTypeHandler()
		});
		sessionFactory.setTypeAliases(new Class[] {
			DateTime.class,
			Role.class,
			Manager.class,
		});
		return sessionFactory;
	}
```

A call to the method findByLogin of my `ManagerRepository.java`

```java
public interface ManagerRepository extends PagingAndSortingRepository<Manager, Long> {
	public Manager findByLogin(String login);
}
```

Result in an error in SqlSessionTemplate throwed in  `SqlSessionInterceptor`.

```
### Error querying database.  Cause: org.apache.ibatis.executor.result.ResultMapException: Error attempting to get column 'roleList' from result set.  Cause: java.lang.IllegalArgumentException: No enum constant com.domain.Role.2
### The error may exist in com.domain.Manager.findByLogina932691fd7d64c599c77cfe05cc71f4d_auto_generate.xml
### The error may involve com.domain.Manager.findByLogina932691fd7d64c599c77cfe05cc71f4d
### The error occurred while handling results
### SQL: select `manager`.id as `id`,`manager`.login as `login`,`manager`.password as `password`,`manager`.role_list as `roleList` from manager `manager`  where   (  `manager`.login=? )
### Cause: org.apache.ibatis.executor.result.ResultMapException: Error attempting to get column 'roleList' from result set.  Cause: java.lang.IllegalArgumentException: No enum constant com.domain.Role.2
```

I have no idea how to fix this and why this is happening in the first place.

This type handler was storing and retrieving Role correctly without installing spring-data-mybatis so I guess it comes from the 1.0.3.

Do you know what's going wrong ?

**Comment by kopax** [comment]
I have replaced my enum role with a list of String[] of roles.

I add to write a very simple typehandler but it doesn't get loaded.

I have this error : 

     Caused by: java.lang.IllegalStateException: Type handler was null on parameter mapping for property 'roles'. It was either not specified and/or could not be found for the javaType ([Ljava.lang.String;) : jdbcType (VARCHAR) combination.

It fails when the XML get builded in `SqlSourceBuilder.java` (L45)

The typeHandler are missing, I think you mean need to do a TypeHandler annotation and insert the typeHandler so it get's available here am I right ? :)

**Comment by easybest** [comment]
I've found the problem, and it will be fixed in 1.0.4

Will be completed tomorrow.

**Comment by easybest** [comment]
this problem will be fixed in #34

**Comment by easybest** [comment]
It's ok now, you can find sample in [https://github.com/hatunet/spring-data-mybatis-samples/tree/issue33](https://github.com/hatunet/spring-data-mybatis-samples/tree/issue33)
